### ISSUE TITLE
Fix build task :generateChangelog

### ISSUE THREAD

**Issue by nipafx** [junit-pioneer/junit-pioneer#730]
I couldn't release 2.0.1 ([build log](https://github.com/junit-pioneer/junit-pioneer/actions/runs/4708049860/jobs/8350319238)) because generating the changelog failed:

```
* What went wrong:
Some problems were found with the configuration of task ':generateChangelog' (type 'GenerateChangelogTask').
  - Gradle detected a problem with the following location: '/home/runner/work/junit-pioneer/junit-pioneer'.
    
    Reason: Task ':generateChangelog' uses this output of task ':compileJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':compileJava' as an input of ':generateChangelog'.
      2. Declare an explicit dependency on ':compileJava' from ':generateChangelog' using Task#dependsOn.
      3. Declare an explicit dependency on ':compileJava' from ':generateChangelog' using Task#mustRunAfter.
    
    Please refer to https://docs.gradle.org/8.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/home/runner/work/junit-pioneer/junit-pioneer'.
    
    Reason: Task ':generateChangelog' uses this output of task ':javadoc' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':javadoc' as an input of ':generateChangelog'.
      2. Declare an explicit dependency on ':javadoc' from ':generateChangelog' using Task#dependsOn.
      3. Declare an explicit dependency on ':javadoc' from ':generateChangelog' using Task#mustRunAfter.
    
    Please refer to https://docs.gradle.org/8.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/home/runner/work/junit-pioneer/junit-pioneer'.
    
    Reason: Task ':generateChangelog' uses this output of task ':javadocJar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':javadocJar' as an input of ':generateChangelog'.
      2. Declare an explicit dependency on ':javadocJar' from ':generateChangelog' using Task#dependsOn.
      3. Declare an explicit dependency on ':javadocJar' from ':generateChangelog' using Task#mustRunAfter.
    
    Please refer to https://docs.gradle.org/8.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/home/runner/work/junit-pioneer/junit-pioneer'.
    
    Reason: Task ':generateChangelog' uses this output of task ':signMavenPublication' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':signMavenPublication' as an input of ':generateChangelog'.
      2. Declare an explicit dependency on ':signMavenPublication' from ':generateChangelog' using Task#dependsOn.
      3. Declare an explicit dependency on ':signMavenPublication' from ':generateChangelog' using Task#mustRunAfter.
    
    Please refer to https://docs.gradle.org/8.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/home/runner/work/junit-pioneer/junit-pioneer'.
    
    Reason: Task ':generateChangelog' uses this output of task ':sourcesJar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':sourcesJar' as an input of ':generateChangelog'.
      2. Declare an explicit dependency on ':sourcesJar' from ':generateChangelog' using Task#dependsOn.
      3. Declare an explicit dependency on ':sourcesJar' from ':generateChangelog' using Task#mustRunAfter.
    
    Please refer to https://docs.gradle.org/8.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
```

Help?

**Comment by Michael1993** [comment]
We could add a `dependsOn(javadoc)` to ensure `generateChangelog` always runs after `javadoc`?

**Comment by nipafx** [comment]
Yesterday I didn't have time to look into it, so I just pasted the error message here. Looking it over now, it appears that the task `:generateChangelog` uses the outputs of tasks `:compileJava`, `:javadoc`, `:javadocJar`, `:signMavenPublication`, and `:sourcesJar` without declaring as much. The [documentation linked in the error message](https://docs.gradle.org/8.1/userguide/validation_problems.html#implicit_dependency) explains why that is a problem and what to do about it.

Ideally, `:generateChangelog` would use tasks or their properties as input, which silently establishes the dependencies on them, but I don't see what we could do to achieve that given that it looks as follows:

```kotlin
generateChangelog {
	val gitFetchRecentTag = Runtime.getRuntime().exec("git describe --tags --abbrev=0")
	val recentTag = gitFetchRecentTag.inputStream.bufferedReader().readText().trim()
	previousRevision = recentTag
	githubToken = System.getenv("GITHUB_TOKEN")
	repository = "junit-pioneer/junit-pioneer"
}
```

It seems to me that all access to Javadoc, JARs, et al are done in the code we don't control. So the only option I see is to add `dependsOn` or `mustRunAfter` on a subset of these graphs. I would assume `:signMavenPublication` would get us a long way as that signs all the JARs and must thus already run after compilation, Javadoc generation, etc.

I will try to reproduce this locally and give this fix a go.

**Comment by nipafx** [comment]
I couldn't fix this in the time I had today. Here's what I tried...

(Before I start, I executed all commands with `./gradlew`, with the option `--console=plain` to keep the full log on the screen, and with the first task `clean` to recreate the clean build environment. I will not put that into every command to keep things readable.)

I looked into `gradle generateChangelog`. As expected it triggers no compilation, no Javadoc, etc. but somewhat unexpected, it works. I was expecting it to fail due to "no Javadoc JAR" or whatever. Then again, it doesn't do anything with the outputs of the other tasks:

```
> Task :generateChangelog
Finding commits between v2.0.0..HEAD in dir: JUnit-Pioneer
Collecting ticket ids from 3 commits.
Fetching ticket info from https://api.github.com/junit-pioneer/junit-pioneer based on 7 ids [725, 728, 729, 726, 727, 713, 716]
Github API rate info => Remaining : 54, Limit : 60, Reset at: 2023-04-16 12:53 PM CEST
Generating changelog based on 7 tickets from Github
Saving changelog to file: JUnit-Pioneer/build/changelog.md
```

So why wouldn't it work? I think it might be worth exploring how to get Gradle to stop complaining about this task.

I also tried to recreate the actual error by running `gradle publishToSonatype closeAndReleaseSonatypeStagingRepository githubRelease` but it doesn't get to changelog generation because it wants to deal with Maven Central first. Indeed, this is what I found in the build log:

```
 > Task :compileJava
> Task :processResources
> Task :classes
> Task :jar
> Task :javadoc
> Task :javadocJar
> Task :sourcesJar
> Task :generateMetadataFileForMavenPublication
> Task :generatePomFileForMavenPublication

> Task :initializeSonatypeStagingRepository
Created staging repository 'orgjunit-pioneer-1048' at https://oss.sonatype.org/service/local/repositories/orgjunit-pioneer-1048/content/

> Task :signMavenPublication
> Task :publishMavenPublicationToSonatypeRepository
> Task :publishToSonatype
> Task :closeSonatypeStagingRepository
> Task :releaseSonatypeStagingRepository
> Task :closeAndReleaseSonatypeStagingRepository
> Task :generateChangelog FAILED
```

Uhm, wait a moment. Doesn't that mean...? Yes, it does! [2.0.1 was actually released.](https://central.sonatype.com/artifact/org.junit-pioneer/junit-pioneer/2.0.1). Nice. I just triggered a website build, so that gets updated and will create a changelog manually after this comment.

But that didn't help me in recreating the problem locally. But as per above message I thought I knew what a fix would be: A dependency on one or more of the previous steps, starting with `:signMavenPublication` because that a very late one. The problem with that is that that task is generated dynamically by the `signing` plugin, so I don't know how to reference it: `dependsOn(signMavenPublication)` simply fails with "Unresolved reference". That's when I gave up...

**Comment by beatngu13** [comment]
>The problem with that is that that task is generated dynamically by the signing plugin, so I don't know how to reference it: dependsOn(signMavenPublication) simply fails with "Unresolved reference". That's when I gave up...

This might work (at least it compiles…):

```diff
diff --git a/build.gradle.kts b/build.gradle.kts
index d9c1cdf..eb76fba 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -358,6 +358,7 @@ tasks {
        }
 
        generateChangelog {
+               dependsOn(nexusPublishing)
                val gitFetchRecentTag = Runtime.getRuntime().exec("git describe --tags --abbrev=0")
                val recentTag = gitFetchRecentTag.inputStream.bufferedReader().readText().trim()
                previousRevision = recentTag
(END)
```

See also:

https://github.com/gradle-nexus/publish-plugin/issues/208

**Comment by beatngu13** [comment]
Looks like we currently pass the extension, not the task:

https://github.com/junit-pioneer/junit-pioneer/actions/runs/6817548891/job/18541551947#step:4:85

**Comment by beatngu13** [comment]
Pioneer 2.2.0 has been successfully released, fixed via 621e808f1954ec7606269b58eccad6fd7df23159.
